ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -1149,6 +1168,22 @@ func (r *RayClusterReconciler) calculateStatus(ctx context.Context, instance *ra
 	// Deep copy the instance, so we don't mutate the original object.
 	newInstance := instance.DeepCopy()
 
+	if features.Enabled(features.RayClusterStatusConditions) {
+		if reconcileErr != nil {
+			if errstd.Is(reconcileErr, reconcilePodsErr) {
+				meta.SetStatusCondition(&newInstance.Status.Conditions, metav1.Condition{
+					Type:    string(rayv1.RayClusterReplicaFailure),
+					Status:  metav1.ConditionTrue,
+					Reason:  "FailedReconcilePods",
```

Comment:
Failing to reconcile pods is different from replica failure, do we need to make this distinction clearer? 

For example, replica failure could mean one of the pods are unhealthy and need to be recreated, but a reconcile error could happen for other reasons before the pod is running

Reply:
Hi @andrewsykim,

Possible distinctions are those in the previous [draft](https://github.com/ray-project/kuberay/pull/2245):
* FailedDeleteAllPods
* FailedDeleteHeadPod
* FailedCreateHeadPod
* FailedDeleteWorkerPod
* FailedCreateWorkerPod

Do they make sense to you?

Reply:
Yes, but my point is that maybe reconcile error and replica failures are better off as separate conditions instead of different reasons within the same condition.

Reply:
We borrow the idea of the replica failures condition from the Kubernetes ReplicaSet controller. It has only two possible reasons:
* FailedCreate
* FailedDelete

And it doesn't reflect the health of pods either. I guess that was because the health of pods is handled by the pod controller already.

The `FailedReconcilePods` reason in this PR may be oversimplified and exceeds the semantic of `ReplicaFailure` indeed. I think I can bring the previous draft back by introducing them as individual error instances and wrapping them at the appropriate points. However, those points are still in the `reconcilePods` function.

``` go
var (
	failedDeleteAllPods   = errstd.New("...")
	failedDeleteHeadPod   = errstd.New("...")
	failedCreateHeadPod   = errstd.New("...")
	failedDeleteWorkerPod = errstd.New("...")
	failedCreateWorkerPod = errstd.New("...")
)

func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv1.RayCluster) error {
		...
		if err := r.createHeadPod(ctx, *instance); err != nil {
			common.FailedClustersCounterInc(instance.Namespace)
			return errstd.Join(failedCreateHeadPod, err)
		}
		...
}
```

Reply:
```go
var (
	failedDeleteAllPods   = errstd.New("...")
	failedDeleteHeadPod   = errstd.New("...")
	failedCreateHeadPod   = errstd.New("...")
	failedDeleteWorkerPod = errstd.New("...")
	failedCreateWorkerPod = errstd.New("...")
)
...
```

This makes sense to me.